src/test/java/com/google/sps/servlets/TestWeatherServlet.java

```diff
@@ -0,0 +1,131 @@
+  @Mock
+  HttpServletResponse response;
+  @Rule
+  public MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  @Test
+  public void testWeatherServlet_doGet_withEmptyLocation_returnsErrorAnswer() throws IOException {
+    when(request.getParameter("location")).thenReturn("");
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    new WeatherServlet(weather, geoCoding).doGet(request, response);
+
+    Assert.assertTrue(stringWriter.toString().equals("Location cannot be null or empty\n"));
+  }
+
+  @Test
+  public void testWeatherServlet_doGet_withLocation_returnsAnswer() throws IOException {
+    when(request.getParameter("location")).thenReturn(RUSSIA_MOSCOW_ADDRESS);
+    when(geoCoding.fromAddressOrPlaceIdToLatLng(RUSSIA_MOSCOW_ADDRESS)).thenReturn(RUSSIA_MOSCOW_LATLNG);
+    when(weather.atLatLngNow(RUSSIA_MOSCOW_LATLNG)).thenReturn(new Weather(10.7, 11.0, 10, 100, 0, "Clear", "10n"));
+    
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    new WeatherServlet(weather, geoCoding).doGet(request, response);
+
+    JsonElement json = JsonParser.parseString(stringWriter.toString());
+    JsonObject jsonObject = json.getAsJsonObject();
+    double temperature = jsonObject.get("temperature").getAsDouble();
+    double temperatureFeelsLike = jsonObject.get("temperatureFeelsLike").getAsDouble();
+    int pressure = jsonObject.get("pressure").getAsInt();
+    int humidity = jsonObject.get("humidity").getAsInt();
+    int clouds = jsonObject.get("clouds").getAsInt();
+    String type = jsonObject.get("type").getAsString();
+    String iconId = jsonObject.get("iconId").getAsString();
+    Assert.assertEquals(10.7, temperature, EPSILON);
+    Assert.assertEquals(11.0, temperatureFeelsLike, EPSILON);
+    Assert.assertEquals(10, pressure);
+    Assert.assertEquals(100, humidity);
+    Assert.assertEquals(0, clouds);
+    Assert.assertEquals("Clear", type);
+    Assert.assertEquals("10n", iconId);
+  }
+
+  @Test
+  public void testWeatherServlet_doGet_withLocationThroughHours_returnsErrorAnswer() throws IOException {
+    when(request.getParameter("location")).thenReturn(RUSSIA_MOSCOW_ADDRESS);
+    when(geoCoding.fromAddressOrPlaceIdToLatLng(RUSSIA_MOSCOW_ADDRESS)).thenReturn(RUSSIA_MOSCOW_LATLNG);
+    when(request.getParameter("hours")).thenReturn("50");
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    new WeatherServlet(weather, geoCoding).doGet(request, response);
+
+    Assert.assertTrue(stringWriter.toString().equals("Hours should be in range 0 <= hours < 48\n"));
+  }
+
+  @Test
+  public void testWeatherServlet_doGet_withLocationThroughDays_returnsErrorAnswer() throws IOException {
```

Comment:
Same as above, make clear why the error happens. However, in this test I don't see any error :)